primer3/primerdesign.pyx

```diff
@@ -0,0 +1,1075 @@
+    reverse_complement_seq_lib(sl)
+    return sl
+
+
+cdef object pdh_design_output_to_dict(
+        const p3_global_settings* global_settings_data,
+        const seq_args* sequence_args_data,
+        const p3retval *retval,
+):
+    cdef:
+        # The pointers to warning tag
+        char* warning = NULL
+
+        # A place to put a string containing all error messages
+        pr_append_str* combined_retval_err = NULL
+
+        # Pointers for the primer set just printing
+        primer_rec* fwd = NULL
+        primer_rec* rev = NULL
+        primer_rec* intl = NULL
+
+        # Variables only used for Primer Lists
+        int num_fwd, num_rev, num_int, num_pair
+        int num_print = 0
+        int print_fwd = 0
+        int print_rev = 0
+        int print_int = 0
+
+        # Switches for printing this primer
+        int go_fwd = 0
+        int go_rev = 0
+        int go_int = 0
+
+        double temp_double = 0
+
+        # The number of loop cycles
+        int loop_max
+
+        # That links to the included region
+        int i
+        int incl_s = sequence_args_data[0].incl_s
+
+        int product_size = 0
+
+        # This deals with the renaming of the internal oligo
+        new_oligo_name = "INTERNAL"
+        int_oligo = new_oligo_name
+
+    output_dict: Dict[str, Any] = {}
+
+    # Check if there are warnings and print them
+    warning = p3_get_rv_and_gs_warnings(retval, global_settings_data)
+    if warning != NULL:
+        warning_b = <bytes> warning
+        output_dict["PRIMER_WARNING"] = warning_b.decode('utf8')
+        free(warning)
+        warning = NULL
+
+
+    combined_retval_err = create_pr_append_str()
+    if combined_retval_err == NULL:
+        raise OSError("Primer3 ran out of memory.")
+
+    try:
+        if pr_append_new_chunk_external(combined_retval_err, retval[0].glob_err.data):
+            raise OSError("Primer3 ran out of memory.")
+
+        # NOTE: These are non fatal errors
+        if pr_append_new_chunk_external(combined_retval_err, retval[0].per_sequence_err.data):
+            raise OSError("Primer3 ran out of memory.")
+
+        # Check if there are errors, print and return
+        if not pr_is_empty(combined_retval_err):
+            err_msg_b = <bytes> pr_append_str_chars(combined_retval_err)
+            raise OSError(err_msg_b.decode('utf8'))
+
+    finally:
+        destroy_pr_append_str(combined_retval_err)
+        combined_retval_err = NULL
+
+    # Get how many primers are in the array
+    num_fwd = retval[0].fwd.num_elem
+    num_rev = retval[0].rev.num_elem
+    num_int = retval[0].intl.num_elem
+    num_pair = retval[0].best_pairs.num_pairs
+
+    # Prints out selection statistics about the primers
+    if (
+        (global_settings_data[0].pick_left_primer == 1) and
+        not (global_settings_data[0].pick_anyway and sequence_args_data[0].left_input)
+    ):
+        explain_str_b = <bytes> p3_get_oligo_array_explain_string(
```

Comment:
Just confirming that we don't need to explicitly manage the memory occupied by this `explain_string` (similar question for the below code)

Reply:
(I may be missing it but I don't see where `retval` is free'd)

Reply:
Oh never mind, found it in `run_design` (`destroy_p3retval(retval)`) -- leaving this here just for a sanity check